src/transactions/TransactionFrame.cpp

```diff
@@ -471,57 +489,64 @@ TransactionFrame::processFeeSeqNum(AbstractLedgerTxn& ltx, int64_t baseFee)
 }
 
 void
-TransactionFrame::removeUsedOneTimeSignerKeys(
-    SignatureChecker& signatureChecker, AbstractLedgerTxn& ltx)
+TransactionFrame::removeOneTimeSignerFromAllSourceAccounts(
+    AbstractLedgerTxn& ltx) const
 {
-    for (auto const& usedAccount : signatureChecker.usedOneTimeSignerKeys())
+    auto header = ltx.loadHeader();
+    auto ledgerVersion = header.current().ledgerVersion;
+
+    if (ledgerVersion == 7)
     {
-        removeUsedOneTimeSignerKeys(ltx, usedAccount.first, usedAccount.second);
+        return;
     }
-}
 
-void
-TransactionFrame::removeUsedOneTimeSignerKeys(
-    AbstractLedgerTxn& ltx, AccountID const& accountID,
-    std::set<SignerKey> const& keys) const
-{
-    auto account = stellar::loadAccount(ltx, accountID);
-    if (!account)
+    std::unordered_set<AccountID> accounts{getSourceID()};
+    for (auto& op : mOperations)
     {
-        return; // probably account was removed due to merge operation
+        accounts.emplace(op->getSourceID());
     }
 
-    auto header = ltx.loadHeader();
-    auto changed = std::accumulate(
-        std::begin(keys), std::end(keys), false,
-        [&](bool r, const SignerKey& signerKey) {
-            return r || removeAccountSigner(header, account, signerKey);
-        });
-
-    if (changed)
+    auto signerKey = SignerKeyUtils::preAuthTxKey(*this);
+    for (auto const& accountID : accounts)
     {
-        normalizeSigners(account);
+        removeAccountSigner(ltx, header, accountID, signerKey);
     }
 }
 
-bool
-TransactionFrame::removeAccountSigner(LedgerTxnHeader const& header,
-                                      LedgerTxnEntry& account,
+void
+TransactionFrame::removeAccountSigner(AbstractLedgerTxn& ltx,
+                                      LedgerTxnHeader const& header,
+                                      AccountID const& accountID,
                                       SignerKey const& signerKey) const
 {
-    auto& acc = account.current().data.account();
+    auto accountNoRecord = stellar::loadAccountWithoutRecord(ltx, accountID);
+    if (!accountNoRecord)
+    {
+        // account was probably removed due to merge operation
+        return;
+    }
+
+    auto& signersNoRecord = accountNoRecord.current().data.account().signers;
     auto it = std::find_if(
-        std::begin(acc.signers), std::end(acc.signers),
+        std::begin(signersNoRecord), std::end(signersNoRecord),
         [&signerKey](Signer const& signer) { return signer.key == signerKey; });
-    if (it != std::end(acc.signers))
+
+    if (it != std::end(signersNoRecord))
     {
+        auto indexToRemove = it - std::begin(signersNoRecord);
+        auto keyToRemove = it->key;
+
+        // The signer exists, so we can load the account to modify it now
+        accountNoRecord.deactivate();
```

Comment:
We've worked really hard to make sure `deactivate` never appears explicitly outside of the Ledger subsystem. This isn't about style. We've taken this approach because `deactivate` exposes the potential for terrible errors. None of these errors seem to exist in your code, but consider some of the things that could go wrong:
- `accountNoRecord` could be accessed directly
- A reference to something in `accountNoRecord` like `signersNoRecord` could be accessed
- An iterator like `it` could be accessed

Typically, we take the approach of deactivating entries by letting them go out of scope. This adds a considerable level of protection against the aforementioned issues (`accountNoRecord` is no longer accessible, ideally `signersNoRecord` and `it` have also gone out of scope and are thus no longer accessible). I'd like to take this approach here as well. One clean way to restructure this would be to do everything related to `accountNoRecord` in a scope at the beginning of this function and use `indexToRemove` to extract the important detail from that scope.